Review followup: PR #22946 — GH#22941: fix: retry failures after recovery releases

## Unaddressed review bot suggestions

PR #22946 was merged with unaddressed review bot feedback. Each comment
below includes its file path, line number, a direct link to the inline
review comment, and a `diff` fence with the code context the bot was
flagging. Resolved and outdated threads are filtered out via GitHub's
GraphQL review-thread state. Read the relevant lines, decide whether
the suggestion is correct, and either apply the fix or close this issue
with a wontfix rationale.

**Source PR:** https://github.com/marcusquinn/aidevops/pull/22946

---

### You are the triager (worker-is-triager rule)

This issue is **auto-created from review bot output** and dispatched
directly to you. Review bots can be wrong: hallucinated line refs, false
premises about codebase structure, template-driven sweeps without
measurements (see GH#17832-17835 for prior art and `AGENTS.md`
"AI-Generated Issue Quality"). **Do not assume the bot is correct.** Verify before acting.

You must end in exactly one of three outcomes — no fourth "hand it back
to the human" path exists. Humans approve decisions; they do not re-do
analysis.

#### Outcome A — Premise falsified → close the issue

1. Read the cited `file:line` (listed under *Files to modify* below).
2. If the bot's claim is factually wrong (file doesn't exist at that
   line, function doesn't behave as described, "auto-generated" section
   isn't actually auto-generated, etc.), **close the issue** with a
   comment in this shape:

   > **Premise falsified.** \<what the bot claimed\>. \<what the code
   > actually shows, with a `file:line` citation or one-line quote\>.
   > Not acting.

   No PR. No further dispatch. The closing comment trains the next
   session reading this thread and the noise filter.

#### Outcome B — Premise correct + fix is obvious → implement and PR

1. Verify the bot's premise as above.
2. Read the Worker Guidance section below, open a worktree, implement.
3. Open a PR with `Resolves #<this-issue-number>` in the body
   (use THIS issue's number, not the source PR's) so merge auto-closes it.
4. Follow the normal Lifecycle Gate (brief, tests, review-bot-gate,
   merge, postflight).

#### Outcome C — Premise correct but approach is a genuine judgment call

Only use this path if you reach it after Outcomes A and B don't apply:
the bot's finding is real, but the fix requires a decision that is
architectural, policy, breaking-change, or otherwise genuinely outside
what you can resolve autonomously. In that case, post a **decision
comment** with exactly these fields:

- **Premise check:** one line, confirming the finding is real.
- **Analysis:** 2-4 bullets on the trade-offs.
- **Recommended path:** the option you would take if the decision were
  yours, with rationale.
- **Specific question:** the single decision the human needs to make
  (yes/no or pick-one, not open-ended).

Then apply `needs-maintainer-review` and stop. The human wakes up to a
ready-to-approve recommendation, not a blank task.

> **Ambiguity about scope or style is not Outcome C.** Per
> `AGENTS.md` "Reasoning responsibility", the model does the
> thinking and delivers a recommendation. Only escalate what is genuinely
> a maintainer-only decision.

### Worker Guidance

**Files to modify:**

- `.agents/scripts/pulse-dispatch-worker-launch.sh:272`
- `.agents/scripts/pulse-fast-fail.sh:21`

**Implementation steps (Outcome B path):**

1. Read the `diff` block under each inline comment below — it shows the
   exact code the bot was flagging. Open the file only if you need
   surrounding context beyond what the diff tail shows.
2. Read the bot's full comment below the diff — it contains the rationale
   and any suggested change.
3. Verify the premise before implementing (see Outcome A). If the premise
   is wrong, switch to Outcome A instead of burning iterations trying to
   satisfy a wrong suggestion.
4. If multiple comments target the same file, group your edits into one
   logical commit.
5. Run `shellcheck` / `markdownlint-cli2` / project tests as appropriate.

**Verification:**

- Open the new PR with `Resolves #<this-issue>` so this followup is auto-closed on merge.
- If the bot's suggestion was incorrect, close this issue with a Outcome A comment — do not open a no-op PR.

### Inline comments

#### gemini-code-assist on `.agents/scripts/pulse-dispatch-worker-launch.sh:272`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/22946#discussion_r3190398304)

```diff
+	[[ -n "$repo_slug" ]] || { printf '0'; return 0; }
+	command -v gh >/dev/null 2>&1 || { printf '0'; return 0; }
+
+	local count=""
+	# shellcheck disable=SC2016  # jq program is intentionally single-quoted.
+	count=$(gh api "repos/${repo_slug}/issues/${issue_number}/comments" --paginate \
+		--jq '[.[] | select((.body // "") | test("worker_noop_zero_output|zero[- ]output|classified as `no_work`"; "i"))] | length' \
+		2>/dev/null) || count="0"
+	[[ "$count" =~ ^[0-9]+$ ]] || count=0
+	printf '%s' "$count"
 	return 0
 }
```

> ![high](https://www.gstatic.com/codereviewagent/high-priority.svg)
> 
> This function is a duplicate of the one already defined at line 274. In Bash, the last definition of a function overrides any previous ones, meaning the implementation added here will never be executed. Furthermore, this new implementation is missing the logic to sum results across multiple pages of comments (the `awk` summation seen in the second definition) and ignores the `ZERO_OUTPUT_COMMENT_EVIDENCE_ENABLED` configuration toggle. You should remove this duplicate definition and instead update the existing function at line 274 with the desired regex improvements. When updating the `awk` logic, ensure it is independently robust against potential errors such as division-by-zero.
> 
> <details>
> <summary>References</summary>
> 
> 1. To improve maintainability, ensure that commands like `awk` are independently robust and handle potential errors such as division-by-zero, even when outer logic already provides a safeguard.
> 2. In shell scripts, extract repeated logic into an internal helper function to improve maintainability.
> </details>

#### gemini-code-assist on `.agents/scripts/pulse-fast-fail.sh:21`

[View inline comment](https://github.com/marcusquinn/aidevops/pull/22946#discussion_r3190398313)

```diff
@@ -16,6 +16,9 @@
 # Functions in this module (in source order):
 #   - _ff_key
 #   - _ff_load
+#   - _ff_current_aidevops_version
+#   - _ff_version_gt
+#   - _ff_entry_is_from_older_release
```

> ![medium](https://www.gstatic.com/codereviewagent/medium-priority.svg)
> 
> The function name `_ff_entry_is_from_older_release` listed in the header does not match the actual implementation in the file, which is named `_ff_release_retry_reset_if_newer` (line 380). Please update the header to reflect the correct function name.
> 
> ```suggestion
> #   - _ff_release_retry_reset_if_newer
> ```

### PR review summaries

#### gemini-code-assist[bot] review summary

[View review](https://github.com/marcusquinn/aidevops/pull/22946#pullrequestreview-4230328025)

> ## Code Review
> 
> This pull request implements GitHub issue comment counting for failure tracking in pulse-dispatch-worker-launch.sh and updates function headers in pulse-fast-fail.sh. The review identifies a duplicate function definition that overrides existing logic while missing pagination and configuration features. Additionally, a naming mismatch was found between the header documentation and the actual function implementation in pulse-fast-fail.sh.
<!-- aidevops:sig -->
---
[aidevops.sh](https://aidevops.sh) v3.14.75 automated scan.

--- a/.agents/scripts/pulse-dispatch-worker-launch.sh
+++ b/.agents/scripts/pulse-dispatch-worker-launch.sh
@@ -263,9 +263,10 @@ _dlw_zero_output_comment_count() {
 	[[ -n "$repo_slug" ]] || { printf '0'; return 0; }
 
 	local count=""
+	# shellcheck disable=SC2016  # jq program is intentionally single-quoted.
 	count=$(gh api --paginate "repos/${repo_slug}/issues/${issue_number}/comments?per_page=100" \
-		--jq '[.[] | select(.body | test("CLAIM_RELEASED reason=worker_noop_zero_output|worker_noop_zero_output|zero[- ]output"; "i"))] | length' 2>/dev/null | \
-		awk '{ total += $1 } END { print total + 0 }') || count=0
+		--jq '[.[] | select((.body // "") | test("CLAIM_RELEASED reason=worker_noop_zero_output|worker_noop_zero_output|zero[- ]output|classified as `no_work`"; "i"))] | length' 2>/dev/null | \
+		awk '{ if ($1 ~ /^[0-9]+$/) { total += $1 } } END { printf "%d", total + 0 }') || count=0
 	[[ "$count" =~ ^[0-9]+$ ]] || count=0
 	printf '%s' "$count"
 	return 0
@@ -280,8 +281,9 @@ _dlw_comment_bloat_metrics() {
 	[[ -n "$repo_slug" ]] || { printf '0\t0\t0\t0'; return 0; }
 
 	local metrics=""
+	# shellcheck disable=SC2016  # jq program is intentionally single-quoted.
 	metrics=$(gh api --paginate "repos/${repo_slug}/issues/${issue_number}/comments?per_page=100" \
-		--jq '[.[] | {body: (.body // "")}] | {comments: length, ops: ([.[] | select(.body | test("ops:start|DISPATCH_CLAIM|CLAIM_RELEASED|dispatch-cooldown|Worker Watchdog Kill"; "i"))] | length), zero: ([.[] | select(.body | test("CLAIM_RELEASED reason=worker_noop_zero_output|worker_noop_zero_output|zero[- ]output"; "i"))] | length), chars: ([.[].body | length] | add // 0)} | [.comments, .ops, .zero, .chars] | @tsv' \
+		--jq '[.[] | {body: (.body // "")}] | {comments: length, ops: ([.[] | select(.body | test("ops:start|DISPATCH_CLAIM|CLAIM_RELEASED|dispatch-cooldown|Worker Watchdog Kill"; "i"))] | length), zero: ([.[] | select(.body | test("CLAIM_RELEASED reason=worker_noop_zero_output|worker_noop_zero_output|zero[- ]output|classified as `no_work`"; "i"))] | length), chars: ([.[].body | length] | add // 0)} | [.comments, .ops, .zero, .chars] | @tsv' \
 		2>/dev/null | awk -F '\t' '{c+=$1; o+=$2; z+=$3; ch+=$4} END {printf "%d\t%d\t%d\t%d", c+0, o+0, z+0, ch+0}') || metrics="0	0	0	0"
 	[[ -n "$metrics" ]] || metrics=$'0\t0\t0\t0'
 	printf '%s' "$metrics"

